test/e2e/rule_test.go

```diff
@@ -62,19 +70,321 @@ func createRuleFiles(t *testing.T, dir string) {
+			m.setLastError(errors.Errorf("invalid auth: %s", req.URL.Path))
+			resp.WriteHeader(http.StatusForbidden)
+			return
+		}
+	}
+
+	b, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		m.setLastError(err)
+		resp.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	var alerts []*model.Alert
+	if err := json.Unmarshal(b, &alerts); err != nil {
+		m.setLastError(err)
+		resp.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	m.mtx.Lock()
+	m.alerts = append(m.alerts, alerts...)
+	m.mtx.Unlock()
+}
+
+func TestRuleAlertmanagerHTTPClient(t *testing.T) {
+	a := newLocalAddresser()
+
+	// Plain HTTP with a prefix.
+	handler1 := newMockAlertmanager("/prefix/api/v1/alerts", "")
+	srv1 := httptest.NewServer(handler1)
+	defer srv1.Close()
+	// HTTPS with authentication.
+	handler2 := newMockAlertmanager("/api/v1/alerts", "secret")
+	srv2 := httptest.NewTLSServer(handler2)
+	defer srv2.Close()
+
+	// Write the server's certificate to disk for the alerting configuration.
+	tlsDir, err := ioutil.TempDir("", "tls")
+	defer os.RemoveAll(tlsDir)
+	testutil.Ok(t, err)
+	var out bytes.Buffer
+	err = pem.Encode(&out, &pem.Block{Type: "CERTIFICATE", Bytes: srv2.TLS.Certificates[0].Certificate[0]})
+	testutil.Ok(t, err)
+	caFile := filepath.Join(tlsDir, "ca.crt")
+	err = ioutil.WriteFile(caFile, out.Bytes(), 0640)
+	testutil.Ok(t, err)
+
+	amCfg := serializeAlertingConfiguration(
+		t,
+		alert.AlertmanagerConfig{
+			StaticAddresses: []string{srv1.Listener.Addr().String()},
+			Scheme:          "http",
+			Timeout:         model.Duration(time.Second),
+			PathPrefix:      "/prefix/",
+		},
+		alert.AlertmanagerConfig{
+			HTTPClientConfig: alert.HTTPClientConfig{
+				TLSConfig: alert.TLSConfig{
+					CAFile: caFile,
+				},
+				BearerToken: "secret",
+			},
+			StaticAddresses: []string{srv2.Listener.Addr().String()},
+			Scheme:          "https",
+			Timeout:         model.Duration(time.Second),
+		},
+	)
+
+	rulesDir, err := ioutil.TempDir("", "rules")
+	defer os.RemoveAll(rulesDir)
+	testutil.Ok(t, err)
+	createRuleFiles(t, rulesDir)
+
+	qAddr := a.New()
+	r := rule(a.New(), a.New(), rulesDir, amCfg, []address{qAddr}, nil)
+	q := querier(qAddr, a.New(), []address{r.GRPC}, nil)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	exit, err := e2eSpinup(t, ctx, q, r)
+	if err != nil {
+		t.Errorf("spinup failed: %v", err)
+		cancel()
+		return
+	}
+
+	defer func() {
+		cancel()
+		<-exit
+	}()
+
+	testutil.Ok(t, runutil.Retry(5*time.Second, ctx.Done(), func() (err error) {
+		select {
+		case <-exit:
+			cancel()
+			return nil
+		default:
+		}
+
+		for i, am := range []*mockAlertmanager{handler1, handler2} {
+			if len(am.Alerts()) == 0 {
+				return errors.Errorf("no alert received from handler%d, last error: %v", i, am.LastError())
+			}
+		}
+
+		return nil
+	}))
+}
+
+func TestRuleAlertmanagerFileSD(t *testing.T) {
+	a := newLocalAddresser()
+
+	am := alertManager(a.New())
```

Comment:
What do you think about this and using alertmanager Mock? I like e2e compatibility check against Alertmanager. I guess it would be too hard to use proper alertmanager in `TestRuleAlertmanagerHTTPClient` as well? (:

Reply:
This makes sense totally, worth to comment maybe? (: 

> Side-note: with the Alertmanager v2 API and its Open API specification, it's even less needed to run a "real" Alertmanager server as you can generate the server code and probably hook into it from the e2e tests.

From API perspective yes (methods, required parameters etc), but it's always nice to have e2e tests against actual implementation. This useful to check against hidden invariants etc.

Reply:
Agreed. It would be worth revisiting this part when we add support for Alertmanager API v2. I tried quickly to hack something with `httputil.ReverseProxy` but I failed short.